internal/sysmonitor/memory_linux.go

```diff
@@ -0,0 +1,266 @@
+	}
+
+	if m, err := readCgroupV1Memory(); err == nil {
+		memoryReaderMu.Lock()
+		memoryReader = readCgroupV1Memory
+		memoryReaderMu.Unlock()
+		return m, nil
+	}
+
+	// Fallback to Host (Bare metal / VM / Unlimited Container)
+	m, err := readSystemMemory()
+	if err != nil {
+		return SystemMemory{}, fmt.Errorf("failed to read system memory from all sources: %w", err)
+	}
+	memoryReaderMu.Lock()
+	memoryReader = readSystemMemory
+	memoryReaderMu.Unlock()
+	return m, nil
+}
+
+func readCgroupV2Memory() (SystemMemory, error) {
+	return readCgroupV2MemoryWithFS(fileSystem)
+}
+
+func readCgroupV2MemoryWithFS(fs FileSystem) (SystemMemory, error) {
+	usage, err := readCgroupValueWithFS(fs, "/sys/fs/cgroup/memory.current")
+	if err != nil {
+		return SystemMemory{}, fmt.Errorf("failed to read cgroup v2 memory usage: %w", err)
+	}
+
+	limit, err := readCgroupValueWithFS(fs, "/sys/fs/cgroup/memory.max")
+	if err != nil {
+		return SystemMemory{}, fmt.Errorf("failed to read cgroup v2 memory limit: %w", err)
+	}
+
+	// Parse memory.stat to find reclaimable memory (inactive_file)
+	inactiveFile, _ := readCgroupStatWithFS(fs, "/sys/fs/cgroup/memory.stat", "inactive_file")
+
+	// Available = (Limit - Usage) + Reclaimable
+	// Note: If Limit - Usage is near zero, the kernel would reclaim inactive_file
+	// Handle case where usage exceeds limit
+	var available uint64
+	if usage > limit {
+		available = inactiveFile // Only reclaimable memory is available
+	} else {
+		available = (limit - usage) + inactiveFile
+	}
+
+	if available > limit {
+		available = limit
+	}
+
+	return SystemMemory{
+		Total:     limit,
+		Available: available,
+	}, nil
+}
+
+func readCgroupV1Memory() (SystemMemory, error) {
+	return readCgroupV1MemoryWithFS(fileSystem)
+}
+
+func readCgroupV1MemoryWithFS(fs FileSystem) (SystemMemory, error) {
+	usage, err := readCgroupValueWithFS(fs, "/sys/fs/cgroup/memory/memory.usage_in_bytes")
+	if err != nil {
+		return SystemMemory{}, fmt.Errorf("failed to read cgroup v1 memory usage: %w", err)
+	}
+
+	limit, err := readCgroupValueWithFS(fs, "/sys/fs/cgroup/memory/memory.limit_in_bytes")
+	if err != nil {
+		return SystemMemory{}, fmt.Errorf("failed to read cgroup v1 memory limit: %w", err)
+	}
+
+	// Check for "unlimited" (random huge number in V1)
+	if limit > (1 << 60) {
+		return SystemMemory{}, os.ErrNotExist
+	}
+
+	// Parse memory.stat for V1
+	inactiveFile, _ := readCgroupStatWithFS(fs, "/sys/fs/cgroup/memory/memory.stat", "total_inactive_file")
+
+	// Handle case where usage exceeds limit
+	var available uint64
+	if usage > limit {
+		available = inactiveFile // Only reclaimable memory is available
+	} else {
+		available = (limit - usage) + inactiveFile
+	}
+
+	if available > limit {
+		available = limit
+	}
+
+	return SystemMemory{
+		Total:     limit,
+		Available: available,
+	}, nil
+}
+
+func readCgroupValueWithFS(fs FileSystem, path string) (uint64, error) {
+	data, err := fs.ReadFile(path)
+	if err != nil {
+		return 0, err
```

Comment:
Improve error handling by wrapping the error with additional context (across the package):
```go
fmt.Errorf("failed to read file %s: %w", path, err)
```